### PR TITLE
Make TAIT behave exactly like RPIT

### LINKED ISSUE #96552: Experiment: try out removing opaque types from typeck expectations
https://github.com/rust-lang/rust/commit/a721052457cf513487fb4266e3ade65c29b272d2 made type-alias-impl-trait different from return-position-impl-trait.

```rust
type Foo = impl std::fmt::Debug;

fn foo(b: bool) -> Foo {
    if b {
        vec![42_i32]
    } else {
        std::iter::empty().collect()
        //~^ ERROR `Foo` cannot be built from an iterator over elements of type `_`
    }
}
```

does not compile while the equivalent return-position-impl-trait program compiles:

```rust
fn foo(b: bool) -> impl std::fmt::Debug{
    if b {
        vec![42_i32]
    } else {
        std::iter::empty().collect()
    }
}
```

We should figure out what we actually want here and whether we can go even further and just remove opaque types from Expectation entirely. Things to try:

* [ ] `fn foo() -> for<'a> impl FnOnce(&'a str) -> usize { |s| s.len() }`
* [ ] `fn foo() -> impl FnOnce(&'static str) -> usize { |s| s.len() }`

### PR DESCRIPTION
**Pull request by oli-obk**
fixes https://github.com/rust-lang/rust/issues/96552

This makes type-alias-impl-trait behave like return-position-impl-trait. Unfortunately it also causes some cases to stop compiling due to "needing type annotations" and makes panicking cause fallback for the hidden type to `()`.

All of these are addressable, but we should probably address them for RPIT and TAIT together

r? @lcnr

### CODE REVIEW

File: compiler/rustc_infer/src/infer/opaque_types.rs [hunk 1]
```diff
@@ -39,35 +39,33 @@ pub struct OpaqueTypeDecl<'tcx> {
...
-    /// lazy TAIT around RPIT handling.
     pub fn replace_opaque_types_with_inference_vars<T: TypeFoldable<'tcx>>(
         &self,
         value: T,
         body_id: HirId,
         span: Span,
         param_env: ty::ParamEnv<'tcx>,
+        // This is a backwards compatibility hack to prevent breaking changes from
+        // lazy TAIT around RPIT handling.
+        only_for_rpit: bool,
```
Comment (lcnr) [inline compiler/rustc_infer/src/infer/opaque_types.rs:50]: can you use a custom enum for that :thinking: seeing `true` or `false` is somewhat confusing

Comment (oli-obk) [reply compiler/rustc_infer/src/infer/opaque_types.rs]: it was in fact so confusing that porting it over to the enum without looking properly caused a bug :D

**Comment by oli-obk** [comment]
Addressed review comment and rebased (causing some diagnostics changes in positions of diagnostics)

**Comment by lcnr** [comment]
as a test, this should still fail with your changes:
```rust
type Foo = impl std::fmt::Debug;

fn foo(b: bool) -> Foo {
    let x: Foo = if b {
        vec![42_i32]
    } else {
        std::iter::empty().collect()
        //~^ ERROR `Foo` cannot be built from an iterator over elements of type `_`
    };
    x
}
```

**Pull request merged**